gobblin-modules/gobblin-kafka-09/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSimpleStreamingExtractor.java

```diff
@@ -0,0 +1,216 @@
+      return _lwm.compareTo(ko._lwm);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null)
+        return false;
+      if (!(obj instanceof KafkaWatermark))
+          return false;
+      return this.compareTo((CheckpointableWatermark)obj) == 0;
+    }
+
+    @Override
+    public int hashCode() {
+      final int prime = 31;
+      return _topicPartition.hashCode() * prime + _lwm.hashCode();
+    }
+
+    public TopicPartition getTopicPartition() {return _topicPartition;};
+
+    public LongWatermark getLwm() {return  _lwm;};
+  }
+
+
+  private final KafkaSchemaRegistry<String, String> _kafkaSchemaRegistry;
+  private Consumer<String, byte[]> _consumer;
+  private TopicPartition _partition;
+  private Iterator<ConsumerRecord<String, byte[]>> _records;
+  long _rowCount = 0;
+
+  public KafkaSimpleStreamingExtractor(WorkUnitState state) {
+    super(state);
+    _kafkaSchemaRegistry = new SimpleKafkaSchemaRegistry(state.getProperties());
+    _consumer = KafkaSimpleStreamingSource.getKafkaConsumer(state);
+    closer.register(_consumer);
+    _partition = new TopicPartition(state.getProp(KafkaSimpleStreamingSource.TOPIC_NAME),
+                                    state.getPropAsInt(KafkaSimpleStreamingSource.PARTITION_ID));
+    _consumer.assign(Collections.singletonList(_partition));
+    OffsetAndMetadata offset = _consumer.committed(_partition);
+    if (offset == null)
+      _consumer.seekToBeginning(_partition);
+    else
+      _consumer.seek(_partition, offset.offset());
+  }
+
+  /**
+  * Get the schema (metadata) of the extracted data records.
+  *
+  * @return the Kafka topic being extracted
+  * @throws IOException if there is problem getting the schema
+  */
+  @Override
+  public String getSchema() throws IOException {
+    try {
+      return this._kafkaSchemaRegistry.getLatestSchemaByTopic(this._partition.topic());
+    } catch (SchemaRegistryException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public List<Tag<?>> generateTags(State state) {
+    List<Tag<?>> tags = super.generateTags(state);
+    tags.add(new Tag<>("kafkaTopic", state.getProp(KafkaSimpleStreamingSource.TOPIC_WHITELIST)));
+    return tags;
+  }
+
+  /**
+   * Return the next record when available. Will never time out since this is a streaming source.
+   * TODO: Add interrupt so Gobblin can shut down the source.
+   */
+  @Override
+  public RecordEnvelope<byte[]> readRecordImpl(RecordEnvelope<byte[]> reuse) throws DataRecordException, IOException {
+    while ((_records == null) || (!_records.hasNext())) {
+      _records = _consumer.poll(100).iterator();
+    }
+    ConsumerRecord<String, byte[]> record = _records.next();
+    _rowCount++;
+    return new RecordEnvelope<byte[]>(record.value(), new KafkaWatermark(_partition, new LongWatermark(record.offset())));
+  }
+
+  @Override
+  public long getExpectedRecordCount() {
+    return _rowCount;
```

Comment:
AFAICT this stuff is checked during data quality checks during publish (expected == extracted == written). @ibuenros : what is the recommended response if you don't know ahead of time what the expected record count is?